keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,282 @@
+
+```yaml
+AA
+└── BB
+```
+Operator `BB` has a required and empty parameter `PASSWORD`. To provide a way for the `AA` operator user to set the password we extend the operator-task with a new field `parameterFile`:
+
+```yaml
+tasks:
+- name: deploy-bb
+  kind: Operator
+  spec:
+    parameterFile: bb-params.yaml # optional, defines the parameter that will be set on the bb-instance  
+```
+
+The contents of the `bb-params.yaml` and the top-level `AA` `params.yaml`:
+
+```yaml
+# operator-aa/templates/bb-params.yaml
+# Note that I placed it under templates mostly because it also uses templating
+PASSWORD: {{ .Params.BB_PASSWORD }}
+```
+
+The `PASSWORD` value is computed on the server-side when IC executes the `deploy-bb` task. The `BB_PASSWORD` parameter is defined as usual in the top-level `params.yaml` file. 
+
+```yaml
+# operator-aa/params.yaml
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: BB_PASSWORD
+    displayName: "BB password"
+    description: "password for the underlying instance of BB"
+    required: true
+```
+
+This is where we see the encapsulation is action. Every operator that incorporates other operators has to define all necessary parameters at the top-level. When installing the operator `AA` the user then has to define the `BB_PASSWORD` as usual:
+
+```bash
+$ kubectl kudo install AA -p BB_PASSWORD=secret
+```  
+
+which will create `OperatorVersion-AA` 
+
+```yaml
+# /apis/kudo.dev/v1beta1/namespaces/default/operatorversions/aa-0.1.0
+spec:
+  operator:
+    kind: Operator
+    name: dummy
+  parameters:
+  - name: BB_PASSWORD
+    displayName: "BB password"
+    description: "password for the underlying instance of BB"
+    required: true
+  tasks:
+  - name: deploy-bb
+    kind: Operator
+    spec:
+      parameterFile: bb-params.yaml
+  templates:
+    bb-params.yaml: |
+      PASSWORD: {{ .Params.BB_PASSWORD }}
+  plans:
+    deploy:
+      ...  
+```
+
+and `Instance-AA` resources
+
+```yaml
+# /apis/kudo.dev/v1beta1/namespaces/default/instances/instance-aa
+spec:
+  parameters:
+    BB_PASSWORD: secret
+```
+
+During the execution of the `deploy-bb` task, the `bb-params.yaml` is expanded the same way we expand templates during the apply-task execution. The `deploy-bb` operator-task then creates the `Instance-BB` resource and saves the expanded parameter `PASSWORD: secret` in it.
+
+What happens if we have a deeper nested operator-tasks tree e.g.:
+```yaml
+AA
+└── BB
+    └── CC
+        └── DD  
+            └── EE
+```
+and it is the low-level `EE` operator that needs the password? It is like the dependency injection through constructor parameters: every higher-level operator has to encapsulate the password parameter so that `AA` has the `BB_PASSWORD`, `BB` the `CC_PASSWORD` and so on.
```

Comment:
This means that (assuming that all operators require a password), that AA needs `BB_PASSWORD`, `CC_PASSWORD`, `DD_PASSWORD` and `EE_PASSWORD`, i.e. it would require a top-level-operator to repeat every parameter that is exposed by any of the dependencies. 

This seems to be not really sustainable long term - we already have operators with > 50 Parameters.

I think there is an argument to be made that not all parameters need to be exposed on the top-level, and it might be good to highlight specific parameters that *must* be set, or to prevent certain parameters of dependencies to be set, but I think it should still be possible to customise most of them.

Additionally, this would require a lot of updates on the top-level operator each time a nested operator adds new parameters.

Reply:
> I think there is an argument to be made that not all parameters need to be exposed on the top-level, and it might be good to highlight specific parameters that must be set, or to prevent certain parameters of dependencies to be set, but I think it should still be possible to customise most of them.

I initially thought so as well, but from the customer feedback so far (@akirillov ) it seems like operator devs lean towards only exposing as little as necessary, rather than allowing users to tweak every little knob seven levels deep in the stack. Ensuring that every combination of 100 parameters works is impossible.

Reply:
I would argue that this depends very much on the top-level operator. In his specific case this might be true, but for an operator that can be scaled from a single-instance to a multi-datacenter deployment a lot more knobs might need to be tunable.

And yes - ensuring that all combinations work is impossible, that is true. It still might be required to adjust one or two of those 100 parameters seven levels deep in the stack.

Reply:
> I would argue that this depends very much on the top-level operator.

Right, my gut feeling is that the more complex and mature the system is, the shorter the paths will be along which a parameter is forwarded between operators unchanged.

Reply:
> And yes - ensuring that all combinations work is impossible, that is true. It still might be required to adjust one or two of those 100 parameters seven levels deep in the stack.

In this case we still have the underlying instances available so the user can always `kudo update --instance` them individually.